geomesa-index-api/src/main/scala/org/locationtech/geomesa/index/index/BaseTieredFeatureIndex.scala

```diff
@@ -0,0 +1,157 @@
+/***********************************************************************
+ * Copyright (c) 2013-2018 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.index.index
+
+import org.geotools.factory.Hints
+import org.locationtech.geomesa.index.api.{FilterStrategy, GeoMesaFeatureIndex, QueryPlan, WrappedFeature}
+import org.locationtech.geomesa.index.geotools.GeoMesaDataStore
+import org.locationtech.geomesa.index.index.IndexKeySpace._
+import org.locationtech.geomesa.index.utils.Explainer
+import org.locationtech.geomesa.utils.index.ByteArrays
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+
+trait BaseTieredFeatureIndex[DS <: GeoMesaDataStore[DS, F, W], F <: WrappedFeature, W, R, C, V, K]
+    extends BaseFeatureIndex[DS, F, W, R, C, V, K] {
+
+  import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType.RichSimpleFeatureType
+
+  /**
+    * Tiered key space beyond the primary one, if any
+    *
+    * @param sft simple feature type
+    * @return
+    */
+  protected def tieredKeySpace(sft: SimpleFeatureType): Option[IndexKeySpace[_, _]]
```

Comment:
Why did this need to be optional?

Reply:
the tiered keyspace available might not support the simple feature type - e.g. we could have an SFT without a geometry, in which case none of the secondary index options would be valid. kind of a fringe case, but...